vllm/model_executor/models/mllama.py

```diff
@@ -0,0 +1,1132 @@
+                    _num_tiles.append(img.shape[0])
+                out_num_tiles.append(_num_tiles)
+
+            return MllamaImagePixelInputs(
+                type="pixel_values",
+                data=out_images,
+                aspect_ratio_ids=out_ar_ids,
+                aspect_ratio_mask=out_ar_mask,
+            )
+
+        if image_embeds is not None:
+            raise NotImplementedError
+
+        raise AssertionError("This line should be unreachable.")
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[torch.Tensor],
+        attn_metadata: AttentionMetadata,
+        **kwargs: object,
+    ) -> Union[Tuple, CausalLMOutputWithPast]:
+        if attn_metadata.num_prefill_tokens > 0 and \
+            attn_metadata.num_decode_tokens > 0:
+            raise ValueError("Chunk prefill not supported")
+        image_inputs = self._parse_and_validate_image_input(**kwargs)
+        if image_inputs is None:
+            cross_attention_mask = None
+            full_text_row_masked_out_mask = (
+                attn_metadata.encoder_seq_lens_tensor != 0).reshape(-1, 1).to(
+                    input_ids.device)
+            cross_attention_states = None
+            skip_cross_attention = max(attn_metadata.encoder_seq_lens) == 0
+        else:
+            # llama's reference implementation runs the vision model on CPU
+            pixel_values = image_inputs['data']
+            aspect_ratio_ids = image_inputs['aspect_ratio_ids']
+            aspect_ratio_mask = image_inputs['aspect_ratio_mask']
+            cross_attention_states = self.vision_model(pixel_values,
+                                                       aspect_ratio_ids,
+                                                       aspect_ratio_mask)
+            cross_attention_states = self.multi_modal_projector(
+                cross_attention_states)
+
+            bsz, _, _, _, image_token_dim = tuple(cross_attention_states.shape)
+            cross_attention_states = cross_attention_states.view(
+                bsz, -1, image_token_dim)
+
+            cross_attention_states_flat = torch.zeros(
+                sum(attn_metadata.encoder_seq_lens),
+                image_token_dim,
+                device=cross_attention_states.device,
+                dtype=cross_attention_states.dtype)
+            start_pos = 0
+            for seq_len, vision_token_in_batch in zip(
+                    attn_metadata.encoder_seq_lens, cross_attention_states):
+                end_pos = start_pos + seq_len
+                cross_attention_states_flat[
+                    start_pos:end_pos] = vision_token_in_batch[:seq_len]
+                start_pos = end_pos
+            cross_attention_states = cross_attention_states_flat
+            cross_attention_mask = None  # TODO
+
+            full_text_row_masked_out_mask = torch.ones(
+                (attn_metadata.num_prefill_tokens, 1), dtype=torch.bool)
```

Comment:
Nit: can you define a separate function for these? Ideally we should keep the forward pass of the main model clean.